pkg/generator/latch.go

```diff
@@ -0,0 +1,47 @@
+package generator
+
+import "sync"
+
+// ProtocolLatch increases the internal counter every time protocol execution
+// starts and decreases the counter every time protocol execution completes.
+// The latch implements Protocol interface and can be registered in the
+// Scheduler.
+//
+// The protocol code using the latch must guarantee that:
+// 1. `Lock()` is always called before `Unlock()`
+// 2. `Unlock()` is eventually called for every `Lock()`.
+//
+// Note that the Unlock() function may panic if the conditions are not met.
+type ProtocolLatch struct {
+	counter uint64
+	mutex   sync.RWMutex
```

Comment:
As an alternative, we can get rid of the mutex and use `atomic.AddUint64` for writes and `atomic.LoadUint64` for reads.

Reply:
Yeah, and I started with it but `atomic.AddUint64(&counter, ^uint64(0))` to decrement felt less readable than just mutex. But I am fine with both options and can change it to `atomic.AddUint64` if you prefer it.

Reply:
The current version works for me as well!